server/server.go

```diff
@@ -105,24 +111,136 @@ type server struct {
 	listenURL    *url.URL
 	logger       logger.Logger
 	requestFuncs []kithttp.RequestFunc
+	router       *mux.Router
 	serviceName  string
 	shutdownOnce sync.Once
 }
 
 func (s *server) Boot() {
 	s.bootOnce.Do(func() {
-		handler := s.NewRouter()
+		// Define our custom not found handler. Here we take care about logging,
+		// metrics and a proper response.
+		s.router.NotFoundHandler = http.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Now the router initialization happens only once during boot, which makes perfect sense.